Support Interactives Authoring and Runtime metadata

pjanik opened the pull request:
I'll open separate PR with changes that are required to send the same data to Portal. But this is already working, I've been testing it and question-interactives are refactored to use these new interfaces: https://github.com/concord-consortium/question-interactives/pull/18

scytacki commented on app/models/interactive_run_state.rb:
```diff
@@ -72,24 +83,59 @@ def portal_hash
       id: answer_id,
-      question_type: "iframe_interactive",
-      question_id: interactive.embeddable_id
+      question_id: interactive.embeddable_id,
```
I was worried about id conflicts with legacy question types. But I see that `embeddable_id` includes the embeddable class name with the id. So for these interactives it will either be `mw_interactive_[id]` or `managed_interactive_[id]`. 

And the answer_id does the same thing. So these ids are independent of the type or question_type.

scytacki commented on lara-typescript/src/interactive-api-client/metadata-types.ts:
```diff
@@ -0,0 +1,75 @@
+   * Interactive can optionally set subtype, so Teacher Report can display different icons
+   * or categorize interactives into subcategories.
+   */
+  subtype?: string;
```
should be questionSubType

pjanik replied:
Done.

pjanik commented on docs/interactive-api-client/interfaces/iauthoringmultiplechoicemetadata.md:
```diff
@@ -13,10 +13,10 @@
-* [type](iauthoringmultiplechoicemetadata.md#type)
+* [questionType](iauthoringmultiplechoicemetadata.md#questiontype)
+* [required](iauthoringmultiplechoicemetadata.md#optional-required)
+* [subtype](iauthoringmultiplechoicemetadata.md#optional-subtype)
```
Done.

pjanik commented on app/models/base_interactive.rb:
```diff
@@ -19,9 +19,30 @@ def portal_hash
+      # Otherwise, the default "iframe_interactive" will be used.
+      type: type,
+      # These properties are defined in IAuthoringMetadataBase:
+      subtype: metadata[:subtype],
```
Done.

pjanik commented on app/models/base_interactive.rb:
```diff
@@ -19,9 +19,30 @@ def portal_hash
+      prompt: metadata[:prompt],
+      # These properties are stored in LARA. They're basic interactive properties. Some of them might be unused by
+      # Report or Portal when interactive pretends to be a basic question type. But these services might be extended
+      # to show both basic question answer and optionally provide iframe report view.
```
I've added it in the last commit. 

We just never used this authored state in question level I think. Now, we still don't use it, as we copy some properties to top-level data (like prompt, required etc.). Once I've added the authored state as you suggested, we could just make assumptions about authored state in the report code instead and read the prompt directly from there. It can make sense once we deal only interactives (no more old-style questions anymore). 

Then, LARA wouldn't have to care about data format and the report code could import all the interfaces from lara-interactive-api NPM package.

scytacki replied:
True except for the LARA report, but hopefully we can replace that with the portal-report.

pjanik replied:
Right. Although I guess the LARA report only uses `report_state` to initialize iframes (at least in its current form), so I don't think it cares about any metadata / authored state format.

pjanik commented:
> I don't think that is strictly necessary, but it would be nice to be able to say:
"If you are pretending to be a multiple choice, then you need to have these fields in your authored state, and your runtime state should include these fields". I'm not sure if that is fully possible with typescript.

I thought about it a bit, but I don't really know how we could implement that. Maybe @dougmartin or @kswenson has some ideas.

I only added this in the question-interactives repo:
https://github.com/concord-consortium/question-interactives/pull/18/commits/2e9ac7ef2ba6979ce9776d19de5e0a0a6584321d

This will force all interactives that use `BaseQuestionApp` to implement one of the metadata interfaces. It worked in practice, I had to update fill in the blank and video interactives beacause of TS compile errors. Developer still can mix open response authoring metadata with multiple-choice runtime metadata, but well, this should be pretty easy to catch. :wink:

> I could imagine other multiple choice like objects will want to store additional fields in their choice array. So perhaps the report_service export code should strip out any fields that it doesn't know about from this array, just like it is stripping out any fields it doesn't know about from the top level of the authoredState.

I've added a code that does it. Actually, I was considering not to strip out any fields from top-level before. Just to have fewer things to update when things change. I think additional fields shouldn't cause any problems. But I added it to be consistent and for clarity.

In the future, I'd consider this: https://github.com/concord-consortium/lara/pull/559#discussion_r439122513
So, LARA could just pass authored state and don't care about anything. Then, Report could just inspect this authored state to get what it needs using TS interfaces from lara-interactive-api lib. But it'd require changes in Portal Report code and would make more sense when we use only interactives.

scytacki commented:
Looks good to me. I don't trust the code coverage stats because we aren't including coverage specs of the lara-typescript folder (as far as I know).
